Feature/trac 8 drawing

rachidcalazans opened the pull request:
Update the way to add and remove subviews for BrushTools (Pencil, Erase and Color). 

The issue is about the color view is always there, only set as hidden. 

Because of that when tries to draw in the coordinates where the color view was added does not work.

Commit:
TRAC-8 Refactor BrushToolView

- refactor way to load and show the tools
- add tests for this toggle tools views

Commit:
TRAC-8 Refactor BrushToolView

- extract GroupPencilSizeView for a new class

Commit:
TRAC-8 Refactor BrushToolView

- extract the GroupEraseSizeView

Commit:
TRAC-8 Refactor GroupSizeViews

- add new view GroupSizeView
- extend this new class in the GroupEraseSizeView and GroupPencilSizeView

Commit:
TRAC-8 Update BrushToolViewUiTests

- rename test func name

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        super.init(frame: frame)
+    }
+    
+    func conf(delegate: GroupSizeContract, bundle: Bundle, rootGroupView: CGFloat, heightBaseToCenter: CGFloat, xBaseToCenter: CGFloat) {
```
conf?

rachidcalazans replied:
Removed

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+    }
+    
+    func conf(delegate: GroupSizeContract, bundle: Bundle, rootGroupView: CGFloat, heightBaseToCenter: CGFloat, xBaseToCenter: CGFloat) {
+        self.delegate = delegate
```
Identation

alexandrehcdc replied:
Indentation

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+    }
+    
+    private func loadGroupPencilSizeButtons() {
+        let buttonSize1   = Button(imageName: "\(buttonSizeImageName)1", imageColor: blackHexColor, selector: #selector(selectSize))
```
What do you think about creating a function like "setButton" ? In this function you would have the changed parameters as parameter, like button tag and size.

rachidcalazans replied:
Nice suggestion. Done.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -473,27 +341,37 @@ public class BrushToolView: UIView {
-        delegate?.changePencilSize(CGFloat(sender.tag))
+    fileprivate func removeSubViews() {
+        for subView in rootView.subviews {
+            if subView.accessibilityIdentifier == GroupPencilSizeView.groupPencilSizeViewIdentifier {
```
Does these if blocks can match in the same interaction? (_e.g._: a subview identifier can be GroupPencilSizeView.groupPencilSizeViewIdentifier and GroupEraseSizeView.groupEraseSizeViewIdentifier at the same time? If not, you should use if/else if blocks

rachidcalazans replied:
Done. Thanks! :D

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        setButtonsAsNotClicked(buttons: groupButtons)
+        
+        // Set the second SizeButton as Default
+        setButtonSizeAsClicked(button: groupButtons[1].uiButton)
```
I know that, in this case, you always have this second element.. But I think that you should treat this to avoid app crashing if it does not exists.

rachidcalazans replied:
Done. Thanks

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+            setButtonSizeAsNotClicked(button: button.uiButton)
+        }
+    }
+    private func setButtonSizeAsClicked(button: UIButton) {
```
Add line here.

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        button.backgroundColor = darkGray
+    }
+    
+    
```
Remove line here.

liuvasconcelos commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        
+        groupView.layer.cornerRadius = groupView.frame.width / 2
+        
+        if addAsSubview {
```
I was taking a look here and you only use this function one time here, and the method is private. You are using this variable as false, so this conditional will never be rechearble.

rachidcalazans replied:
Removed. Thanks

liuvasconcelos replied:
Chid, you removed the conditional.. But you are still adding the subview. My question is: you should do it?
Because on previous implementation, you had that conditional and you passed it as false.

rachidcalazans replied:
Sorry. Removed =D

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        groupSizeView.addSubview(separatorView)
+        groupSizeView.addSubview(groupButtonsView)
+        
+        centrilizeVerticalView(view: groupSizeView, heightBaseToCenter: heightBaseToCenter, xBaseToCenter: xBaseToCenter)
```
centrilizeVerticalView -> centralizeVerticalView

rachidcalazans replied:
Thanks! :D

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        groupSizeButtons.append(buttonSize6)
+    }
+    
+    private func configureGroupSizeLayout(groupButtons: [Button], groupSizeView: UIView, iconReferenceName: String, heightBaseToCenter: CGFloat, xBaseToCenter: CGFloat) {
```
Thank you for adding the comments above the lines, helped me to understand it :)

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        button.layer.cornerRadius = button.frame.height / 2
+        button.backgroundColor = darkGray
+    }
+    
```
Indentation on these functions (setButtonSizeAsClicked and setButtonSizeAsNotClicked)

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+        
+        let identifierName = "\(iconReferenceName)SizeReferenceIdentifier"
+        setIdentifierForView(view: iconImageView, identifierName: identifierName)
+        iconImageView.frame                   = frameIconImage
```
Indentation

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    private func configureGroupLayout(buttons: [Button], groupView: UIView, groupViewOirigin: CGPoint, addAsSubview: Bool = true) {
+        let _: [Button] = buttons.enumerated().map() { (index, button) in
```
Why are you using a map if you are not expecting a return value?

rachidcalazans replied:
Changed to use a for.

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+            return CGRect(x: x, y: y, width: buttonWidth, height: buttonHeight)
+        }
+        
+        let x = (groupViewWidth - size.width) / 2
```
Indentation

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -0,0 +1,287 @@
+    private func setButtonImage(button: UIButton, imageName: String, imageColor: String) {
+        let selectColorIcon      = UIImage(named: imageName, in: bundle, compatibleWith: nil)
+        let selectColorImageView = UIImageView()
+        selectColorImageView.image = selectColorIcon
```
Indentation

Commit:
TRAC-8 Update GroupSizeView

- refactor some functions
- adjust some identations

alexandrehcdc commented on StantDrawing/BrushTool/GroupSizeView.swift:
```diff
@@ -217,8 +214,6 @@ public class GroupSizeView: UIView {
-            return button
         }
         
         let lastUiButton = buttons.last!.uiButton
```
Avoid forcing optionals :)

rachidcalazans replied:
Done

Commit:
TRAC-8 Update GroupSizeView

- remove unused addSubView

Commit:
TRAC-8 Update GroupSizeView

- update configureGroupLayout function
- add clause guard for optional attribute
